zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/ZkClientType.java

```diff
@@ -25,18 +25,24 @@
    * creation, callback functionality, and session management. But note that this is more
    * resource-heavy since it creates a dedicated ZK connection so should be used sparingly only
    * when the aforementioned features are needed.
+   *
+   * Valid on SINGLE_REALM only.
    */
   DEDICATED,
```

Comment:
In this case, shall we merge the realm option into the enum item's fields? Then we don't need to check since the user cannot configure this option.

Reply:
@jiajunwang 
The user does have the ability to select the ZkClientType. Before ZooScalability, we allowed users to select the ZkClientType (Dedicated or Shared) when creating a BaseDataAccessor. We are doing this to keep things backward-compatible. See the following examples (copy-pasted from the test cases added as part of this PR) below:

 ```
   // Create ZkBaseDataAccessor, multi-realm, dedicated ZkClient (should fail)
    try {
      accessor = zkBaseDataAccessorBuilder.setRealmMode(RealmAwareZkClient.RealmMode.MULTI_REALM)
          .setZkClientType(ZkClientType.DEDICATED).setZkAddress(firstZkAddress)
          .setRealmAwareZkConnectionConfig(connectionConfigBuilder.build()).build();
      Assert.fail("MULTI_REALM and DEDICATED ZkClientType are an invalid combination!");
    } catch (HelixException e) {
      // Expected
    }

     // Create ZkBaseDataAccessor, multi-realm, shared ZkClient (should fail)
    try {
      accessor = zkBaseDataAccessorBuilder.setRealmMode(RealmAwareZkClient.RealmMode.MULTI_REALM)
          .setZkClientType(ZkClientType.SHARED).setZkAddress(firstZkAddress)
          .setRealmAwareZkConnectionConfig(connectionConfigBuilder.build()).build();
      Assert.fail("MULTI_REALM and SHARED ZkClientType are an invalid combination!");
    } catch (HelixException e) {
      // Expected
    }

     // Create ZkBaseDataAccessor, multi-realm, federated ZkClient, ZkAddress set (should fail)
    try {
      accessor = zkBaseDataAccessorBuilder.setRealmMode(RealmAwareZkClient.RealmMode.MULTI_REALM)
          .setZkClientType(ZkClientType.FEDERATED).setZkAddress(firstZkAddress)
          .setRealmAwareZkConnectionConfig(connectionConfigBuilder.build()).build();
      Assert.fail("MULTI_REALM and FEDERATED ZkClientType do not connect to one ZK!");
    } catch (HelixException e) {
      // Expected
    }


```

Reply:
So to answer your question, because the user may want to choose which ZkClientType to use, it's appropriate that we leave this enum as a separate public enum.

Reply:
> The user does have the ability to select the ZkClientType.

Sure. My question is that once a user selects a certain type, do they have to specify a RealmMode in addition? Since based on your comment, it seems to be pure overhead. Because if they select DEDICATED, then the real mode has to be SINGLE_REALM.
I know this is for backward compatibility, so shall we just fill the RealmMode for the user when they configure the client using ZkClientType? In this case, we don't need the validation logic.

Reply:
@jiajunwang 

That is actually something I've considered and the original implementation was somewhat doing that. But I realized that this type of "backfill" or doing this under the surface was more confusing to the user and adds more difficulty in the usage of the Builder API. We should make sure we require the users to explicitly set the types/configs, otherwise we run the risk of doing things without the user knowing and confusing them.

Reply:
So I have several questions after reviewing more code. Maybe you would like to clarify the plan?

1. The default builder build() method can actually create a client with MultiRealm enabled for DEDICATED client. However, we disable this option by enforcing the validation logic. Was it intentionally done?

2. If this is the case, then the actual build() code can be simplified, right? Since half of the branches won't be called forever.

3. 
Regarding what you replied,

> more confusing to the user and adds more difficulty in the usage of the Builder API

I guess what you meant is that if there are 2 ways of configuring the client and these 2 options are all available in one builder, it would be confusing that which one to use.

I think we discussed this topic before. To make it clearer, my suggestion would be,
1. The new builder does not allow ZkClientType config. Assume this is for backward compatibility. So we should not introduce to the new APIs.
2. For the older APIs (deprecated ones), as long as the type is passed in, we auto-fill the RealmMode based on the type. Since you mentioned in the comment that the mapping is one to one, this is doable.
3. In this case, there won't be a confusing part, IMO. And we won't need these validation logics everywhere. This helps with future maintenance work greatly.

Not sure if you agree, please feel free to schedule a discussion. Maybe I missed some usage.

BTW, if you are concerned about the already published version, then we can deprecate the ZkClientType setting method in the builders to suggest users not using it for now. The validation part can be left in the code until we make the API clean later.

Reply:
1. I think you're referring to 

```
      case MULTI_REALM:
        try {
          if (_zkClientType == ZkClientType.DEDICATED) {
            // Use a realm-aware dedicated zk client
            zkClient = DedicatedZkClientFactory.getInstance()
                .buildZkClient(connectionConfig, clientConfig);
          } else if (_zkClientType == ZkClientType.SHARED) {
            // Use a realm-aware shared zk client
            zkClient =
                SharedZkClientFactory.getInstance().buildZkClient(connectionConfig, clientConfig);
          } else {
            zkClient = new FederatedZkClient(connectionConfig, clientConfig);
          }
        } catch (IOException | InvalidRoutingDataException | IllegalStateException e) {
          throw new HelixException("Not able to connect on multi-realm mode.", e);
        }
        break;
```

Reviewing this logic again, I believe this was a bug. Multi-realm means it has to use FederatedZkClient only because it has to be able to talk to multiple ZKs. I've fixed this. Good catch!

2. I've removed some code in the createZkClient() method.

3. We can still support ZkClientType after fixing the bug. Like you said, ZkClientType is not set, we default to SHARED (original behavior).

Reply:
Given the current situation that user still has an option to choose between dedicated/shared clients when the realm mode is single realm, I agree that we still need to keep the methods in the builder.